src/watch/index.ts

```diff
@@ -243,31 +224,43 @@ export class Task {
 					result
 				});
 			})
-			.catch((error: Error) => {
+			.catch((error: RollupError) => {
 				if (this.closed) return;
 
-				if (this.cache) {
-					// this is necessary to ensure that any 'renamed' files
-					// continue to be watched following an error
-					if (this.cache.modules) {
-						this.cache.modules.forEach(module => {
-							if (module.transformDependencies) {
-								module.transformDependencies.forEach(depId => {
-									this.watchFile(depId, true);
-								});
-							}
-						});
-					}
-					this.watchFiles.forEach(id => {
+				if (Array.isArray(error.watchFiles)) {
+					for (const id of error.watchFiles) {
 						this.watchFile(id);
-					});
+					}
 				}
 				throw error;
 			});
 	}
 
-	watchFile(id: string, isTransformDependency = false) {
+	private updateWatchedFiles(result: RollupBuild) {
+		const previouslyWatched = this.watched;
+		this.watched = new Set();
+		this.watchFiles = result.watchFiles;
+		this.cache = result.cache;
+		for (const id of this.watchFiles) {
+			this.watchFile(id);
+		}
+		if (this.cache.modules) {
+			for (const module of this.cache.modules) {
+				if (module.transformDependencies) {
+					for (const depId of module.transformDependencies) {
+						this.watchFile(depId, true);
+					}
+				}
+			}
+		}
+		for (const id of previouslyWatched) {
+			if (!this.watched.has(id)) deleteTask(id, this, this.chokidarOptionsHash);
+		}
+	}
+
+	private watchFile(id: string, isTransformDependency = false) {
 		if (!this.filter(id)) return;
+		this.watched.add(id);
```

Comment:
By doing this here, `this.watched` is strictly coupled to the watched files and we do not forget to clean up any of them. Also, `this.watched` should never be reset unless it is made sure that there are no remaining dangling watchers that are no longer tracked.